addons/product/models/product_template.py

```diff
@@ -938,6 +939,83 @@ def _get_first_possible_combination(self, parent_combination=None, necessary_val
+            return
+
+        all_exclusions = self._get_own_attribute_exclusions()
+        # The following dict uses product template attribute values as keys
+        # 0 means the value is acceptable, greater than 0 means it's rejected, it cannot be negative
+        # Bear in mind that several values can reject the same value and the latter can only be included in the
+        #  considered combination if no value rejects it.
+        # This dictionary counts how many times each value is rejected.
+        # Each time a value is included in the considered combination, the values it rejects are incremented
+        # When a value is discarded from the considered combination, the values it rejects are decremented
+        current_exclusions = defaultdict(int)
+        for exclusion in self._get_parent_attribute_exclusions(parent_combination):
+            current_exclusions[exclusion] += 1
+        included_values_ids = set()
+
+        # -1 means no value has been picked for the attribute in the current (partial) combination
+        values_idx = [-1] * len(product_template_attribute_values)
+        # determines which attribute line we're working on
+        attribute_idx = 0
+
+        def current_value_id():
+            return product_template_attribute_values[attribute_idx][values_idx[attribute_idx]].id
+
+        while True:
+            if values_idx[attribute_idx] >= 0:
+                for value_to_include_back in all_exclusions[current_value_id()]:
+                    current_exclusions[value_to_include_back] -= 1
+                included_values_ids.discard(current_value_id())
+
+            # first we try to get to next value of current attribute
+            if values_idx[attribute_idx] < len(product_template_attribute_values[attribute_idx]) - 1:
+                # get to next value of current attribute line
+                values_idx[attribute_idx] += 1
+
+                if current_exclusions[current_value_id()] or \
+                        any(value_to_exclude in included_values_ids for value_to_exclude
+                            in all_exclusions[current_value_id()]):
+                    continue
+
+                for value_to_exclude in all_exclusions[current_value_id()]:
+                    current_exclusions[value_to_exclude] += 1
+                included_values_ids.add(current_value_id())
+            elif attribute_idx != 0:
+                # reset current attribute and get to previous attribute
+                values_idx[attribute_idx] = - 1
+                attribute_idx -= 1
+                continue
+            else:
+                # we're done if we must reset first attribute
+                break
+
+            # then we get to the next attribute or submit combination if we're on the last attribute
+            if attribute_idx == len(product_template_attribute_values) - 1:
+                yield [product_template_attribute_values[pool_idx][value_idx] for pool_idx, value_idx in
```

Comment:
rename pool, though it's not very important since it's a local iteration variable

Reply:
done